### ISSUE TITLE
Placeholder options in config file

### ISSUE THREAD

**Issue by TotalGriffLock** [threatpatrols/opnsense-plugin-autossh#6]
Have just downloaded and installed this plugin on a fresh installation of OPNsense 22.7.6-amd64

The tunnel was created and worked initially, until the OPNsense device was restarted. Now attempts to reconnect fail due to errors in the config file. From the autossh plugin log:

<html>
<body>

2022-10-20T10:07:58 | Informational | autossh | ssh exited with error status 255; restarting ssh |  
-- | -- | -- | -- | --
2022-10-20T10:07:58 | Debug | autossh | check on child 56524 |  
2022-10-20T10:07:58 | Critical | /usr/local/bin/ssh | fatal: /usr/local/etc/autossh/autossh.conf: terminating, 6 bad configuration options |  
2022-10-20T10:07:58 | Error | /usr/local/bin/ssh | error: /usr/local/etc/autossh/autossh.conf line 59: Bad number 'rekey_limit_01': Invalid argument |  
2022-10-20T10:07:58 | Error | /usr/local/bin/ssh | error: /usr/local/etc/autossh/autossh.conf line 57: Bad key types 'pubkey_accepted_key_types_01,pubkey_accepted_key_types_02,pubkey_accepted_key_types_03,pubkey_accepted_key_types_04,pubkey_accepted_key_types_05,pubkey_accepted_key_types_06,pubkey_accepted_key_types_07,pubkey_accepted_key_types_08,pubkey_accepted_key_types_09,pubkey_accepted_key_types_10'. |  
2022-10-20T10:07:58 | Error | /usr/local/bin/ssh | error: /usr/local/etc/autossh/autossh.conf line 48: Bad SSH2 MAC spec 'macs_01,macs_02,macs_03,macs_04,macs_05,macs_06,macs_07,macs_08,macs_09,macs_10'. |  
2022-10-20T10:07:58 | Error | /usr/local/bin/ssh | error: /usr/local/etc/autossh/autossh.conf line 45: Bad SSH2 KexAlgorithms 'kex_algorithms_01,kex_algorithms_02,kex_algorithms_03,kex_algorithms_04,kex_algorithms_05,kex_algorithms_06,kex_algorithms_07,kex_algorithms_08,kex_algorithms_09,kex_algorithms_10,kex_algorithms_11'. |  
2022-10-20T10:07:58 | Error | /usr/local/bin/ssh | error: Unsupported KEX algorithm "kex_algorithms_01" |  
2022-10-20T10:07:58 | Error | /usr/local/bin/ssh | error: /usr/local/etc/autossh/autossh.conf line 37: Bad key types 'host_key_algorithms_01,host_key_algorithms_02,host_key_algorithms_03,host_key_algorithms_04,host_key_algorithms_05,host_key_algorithms_06,host_key_algorithms_07,host_key_algorithms_08,host_key_algorithms_09,host_key_algorithms_10'. |  
2022-10-20T10:07:58 | Error | /usr/local/bin/ssh | error: /usr/local/etc/autossh/autossh.conf line 19: Bad SSH2 cipher spec 'ciphers_01,ciphers_02,ciphers_03,ciphers_04,ciphers_05,ciphers_06'.

</body>
</html>

When checking the autossh.conf file it looks as you would expect from those errors:
```
    HostKeyAlgorithms host_key_algorithms_01,host_key_algorithms_02,host_key_algorithms_03,host_key_algorithms_04,host_key_algorithms_05,host_key_algorithms_06,host_key_algorithms_07,host_key_algorithms_08,host_key_algorithms_09,host_key_algorithms_10
    KexAlgorithms kex_algorithms_01,kex_algorithms_02,kex_algorithms_03,kex_algorithms_04,kex_algorithms_05,kex_algorithms_06,kex_algorithms_07,kex_algorithms_08,kex_algorithms_09,kex_algorithms_10,kex_algorithms_11
    PubkeyAcceptedKeyTypes pubkey_accepted_key_types_01,pubkey_accepted_key_types_02,pubkey_accepted_key_types_03,pubkey_accepted_key_types_04,pubkey_accepted_key_types_05,pubkey_accepted_key_types_06,pubkey_accepted_key_types_07,pubkey_accepted_key_types_08,pubkey_accepted_key_types_09,pubkey_accepted_key_types_10
```

etc etc. It seems that when creating or editing a tunnel from the web interface the config file is generated correctly but when created by configd using the templates it is not

**Comment by TotalGriffLock** [comment]
Further investigation, running this fixes the config file:

root@OPNsense:/usr/local/opnsense/scripts/ThreatPatrols/Autossh # ./autossh.py --debug config_helper

**Comment by ndejong** [comment]
Hi @TotalGriffLock - thanks for the report, the hint that's interesting is your comment 
> ...until the OPNsense device was restarted...

As a work-around we could simply re-invoke the `autossh.py config_helper` at system reboot time, but I'd rather uncover the root-cause - please do comment if you uncover anything else in the meantime.

**Comment by TotalGriffLock** [comment]
Getting there...

```
root@OPNsense:/usr/local # configctl autossh config_helper
{
  "message": "Updated 1 configuration host sections",
  "status": "success"
}
root@OPNsense:/usr/local # grep PubkeyAcceptedKeyTypes etc/autossh/autossh.conf
    PubkeyAcceptedKeyTypes ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ssh-ed25519-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,ssh-ed25519,ssh-rsa
root@OPNsense:/usr/local # configctl template reload ThreatPatrols/Autossh
OK
root@OPNsense:/usr/local # grep PubkeyAcceptedKeyTypes etc/autossh/autossh.conf
    PubkeyAcceptedKeyTypes pubkey_accepted_key_types_01,pubkey_accepted_key_types_02,pubkey_accepted_key_types_03,pubkey_accepted_key_types_04,pubkey_accepted_key_types_05,pubkey_accepted_key_types_06,pubkey_accepted_key_types_07,pubkey_accepted_key_types_08,pubkey_accepted_key_types_09,pubkey_accepted_key_types_10
```
The generation of the config file by the Jinja2 template is where the issue lies, using the "tunnel.pubkey_accepted_key_types" variable is copying the value from config.xml verbatim rather than substituting it with the text from the XML model

**Comment by TotalGriffLock** [comment]
Sorted it!

Update the model XML to include a value="" for each option. Without it the XML tag name is saved rather than the contents. For example:
```
                <pubkey_accepted_key_types type="OptionField">
                    <required>Y</required>
                    <multiple>Y</multiple>
                    <default>pubkey_accepted_key_types_01,pubkey_accepted_key_types_02,pubkey_accepted_key_types_03,pubkey_accepted_key_types_04,pubkey_accepted_key_types_05,pubkey_accepted_key_types_06,pubkey_accepted_key_types_07,pubkey_accepted_key_types_08,pubkey_accepted_key_types_09,pubkey_accepted_key_types_10</default>
                    <optionvalues>
                        <pubkey_accepted_key_types_01 value="ecdsa-sha2-nistp256-cert-v01@openssh.com">ecdsa-sha2-nistp256-cert-v01@openssh.com</pubkey_accepted_key_types_01>
                        <pubkey_accepted_key_types_02 value="ecdsa-sha2-nistp384-cert-v01@openssh.com">ecdsa-sha2-nistp384-cert-v01@openssh.com</pubkey_accepted_key_types_02>
                        <pubkey_accepted_key_types_03 value="ecdsa-sha2-nistp521-cert-v01@openssh.com">ecdsa-sha2-nistp521-cert-v01@openssh.com</pubkey_accepted_key_types_03>
```

Note this seemed to break the config for my existing tunnel, upon editing the model it had no public key types selected and the form would not validate and submit without clicking Advanced and selecting all of the options.

**Comment by TotalGriffLock** [comment]
Same issue with bind_interface field... when written by the web interface it says (e.g.) em1 but on reboot it is recreated with 'lan'

FYI I am putting together a PR to resolve all this but I'm just trying to get my head around the model migration process so users don't have to delete and recreate their tunnels.

**Comment by ndejong** [comment]
For the sake of making this more easily visible:-

As @TotalGriffLock indicates in https://github.com/threatpatrols/opnsense-plugin-autossh/pull/7#issuecomment-1286544530 it is possible to recreate this issue with 

```
configctl template reload ThreatPatrols/Autossh
```

This makes plenty more sense to me now

**Comment by ndejong** [comment]
Merged and new `os-autossh-0.2.42.pkg` package has been pushed to the **stable** repo.

Something I did notice in testing is that if you end up in the situation where `/usr/local/etc/autossh/autossh.conf` is malformed and then you apply this new version, then the `autossh.conf` does not automatically get updated after the install as you might expect - I expect this is because nothing invokes a template reload when the package install occurs, which would otherwise trigger the model migration that you've very nicely provided.

To work-around this situation it's easy enough - just disable and re-enable the tunnel(s) affected - doing so invokes the model migration and everything is happy.

Has been tested on OPNsense 21.1 through to current.

Thanks again @TotalGriffLock this is an important stability fix!
